atomicapp/providers/marathon.py

```diff
@@ -0,0 +1,122 @@
+"""
+ Copyright 2015 Red Hat, Inc.
+
+ This file is part of Atomic App.
+
+ Atomic App is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Lesser General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ Atomic App is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Lesser General Public License for more details.
+
+ You should have received a copy of the GNU Lesser General Public License
+ along with Atomic App. If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import anymarkup
+import requests
+import urlparse
+import logging
+import os
+from atomicapp.plugin import Provider, ProviderFailedException
+from atomicapp.utils import printErrorStatus
+from atomicapp.constants import PROVIDER_API_KEY
+
+logger = logging.getLogger(__name__)
+
+
+class Marathon(Provider):
+    """ Marathon provider
+    """
+
+    key = "marathon"
+    config_file = None
+    marathon_api_version = "v2"
+    # use localhost as default, when no providerurl is specified
+    marathon_api = "http://localhost:8080/%s/" % marathon_api_version
+    marathon_artifacts = []
+
+    def init(self):
+        self.marathon_artifacts = []
+
+        logger.debug("Given config: %s", self.config)
+        if self.config.get(PROVIDER_API_KEY):
+            self.marathon_api = self.config.get(PROVIDER_API_KEY)
+            self.marathon_api = urlparse.urljoin(self.marathon_api, "v2/")
+
+        logger.debug("marathon_api = %s", self.marathon_api)
+        self._process_artifacts()
+
+    def deploy(self):
+        """ Deploy containers
+        """
+        for artifact in self.marathon_artifacts:
+            url = urlparse.urljoin(self.marathon_api, "apps/")
+
+            if self.dryrun:
+                logger.info("DRY-RUN: %s", url)
+                continue
+
+            logger.debug("Deploying appid: %s", artifact["id"])
+            res = requests.post(url, json=artifact)
+            if res.status_code == 201:
+                logger.info(
+                    "Marathon app %s sucessfully deployed.",
+                    artifact["id"])
+            else:
+                msg = "Error deploying app: %s, Marathon API response %s - %s" % (
+                    artifact["id"], res.status_code, res.text)
+                logger.error(msg)
+                raise ProviderFailedException(msg)
+
+    def undeploy(self):
+        """ Undeploy containers
+        """
+        for artifact in self.marathon_artifacts:
+            url = urlparse.urljoin(
+                self.marathon_api,
+                "apps/%s" %
+                artifact["id"])
+
+            if self.dryrun:
+                logger.info("DRY-RUN: %s", url)
+                continue
+
+            logger.debug("Deleting appid: %s", artifact["id"])
```

Comment:
We should _not_ be doing delete at the moment. Currently `undeploy` simply stops the container. Is there the functionality in Mesos to do stop rather than delete?

We will be adding `uninstall` later or something similar that will actually delete the containers on each provider (docker/kubenetes/openshift/etc.)